### PR TITLE
bump FluxC to support the Portfolio PostType.

### PR DESCRIPTION
**Pull request by glung**
/!\ 

Based on https://github.com/wordpress-mobile/WordPress-Android/pull/8221. 

Because the branch backing this PR does not exist in the [WordPress-mobile repo](https://github.com/wordpress-mobile/WordPress-Android ) (it lives in my [fork](https://github.com/glung/WordPress-Android)), I am not able to ask to GitHub to base the PR on this branch.

 **Please, review only the last commit https://github.com/wordpress-mobile/WordPress-Android/pull/8270/commits/093d5d90fdcb05b28760172f5f0c00a95db1ba6c.** 

# Scope 

Bump FluxC with Portfolio support.

This version comes with a breaking API change: the **post type is not represented as a boolean anymore (`isPage)` but as a `PostType`** in order to support Portfolio (please, find some details in the PR https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/869)

This PR aims to migrate the WordPress-Android app to complies with the new API.

# Note for reviewers. 

As aforementionned, please review only the last commit. The rest comes from this other PR: https://github.com/wordpress-mobile/WordPress-Android/pull/8221.

Don't be afraid by the number of changed lines : the logic should not change in any case! 

- Tip to review: any occurence of `isPage ? page : post` is replaced by `getResourceId(isPage, page, post)`
- The `PostType` is an enum, hence serializable and can be passed as it is to intents. 

# Important note! 

Locally, I am not able to build an APK as I consistently experience the following error: 

```
Execution failed for task ':WordPress:transformClassesWithMultidexlistForVanillaDebug'.
> com.android.build.api.transform.TransformException: Error while generating the main dex list.
```

The issue seems to be:
```
com.android.build.api.transform.TransformException: Error while generating the main dex list.
> Error while generating the main dex list.
  > com.android.tools.r8.errors.CompilationError: Program type already present: com.android.volley.Cache
    > Program type already present: com.android.volley.Cache
```

I’ve stop the gradle daemon but also ran a `git clean -fdx` and rebuilt the app. Though, the issue is still here. 

Let's see what happens on CI.

### CODE REVIEW

File: WordPress/build.gradle [hunk 1]
```diff
@@ -174,7 +174,8 @@ dependencies {
     implementation 'com.google.dagger:dagger-android-support:2.11'
     kapt 'com.google.dagger:dagger-android-processor:2.11'
 
-    implementation ("com.github.wordpress-mobile.WordPress-FluxC-Android:fluxc:$fluxCVersion") {
+    // FIXME com.github.wordpress-mobile.WordPress-FluxC-Android
+    implementation ("com.github.glung.WordPress-FluxC-Android:fluxc:$fluxCVersion") {
```
Comment (glung) [inline WordPress/build.gradle]: Update this before merging.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java [hunk 2]
```diff
@@ -63,17 +64,7 @@ protected void onNewIntent(Intent intent) {
     }
 
     private void handleIntent(Intent intent) {
-        mIsPage = intent.getBooleanExtra(EXTRA_VIEW_PAGES, false);
```
Comment (glung) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java]: My understanding is that `pageHasChanged` was always false since `mIsPage` was initialized just before the comparison.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java [hunk 3]
```diff
@@ -102,16 +94,18 @@ private void handleIntent(Intent intent) {
...
+                String errorMessage = getString(PostTypeUtilsKt.getResourceId(
+                        mPostType,
+                        R.string.error_page_does_not_exist,
+                        R.string.error_post_does_not_exist));
                 ToastUtils.showToast(this, errorMessage);
             }
         }
 
         mPostList = (PostsListFragment) getSupportFragmentManager().findFragmentByTag(PostsListFragment.TAG);
-        if (mPostList == null || siteHasChanged || pageHasChanged || targetPost != null) {
```
Comment (glung) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.java]: Cf comment on line 66 above. `pageHasChanged` was always `false`.

**Comment by mzorz** [comment]
Hi @glung 👋  - I was able to run the PR by creating new branches off your working branches on each of the repos (FluxC and WordPressAndroid).

- FluxC repo: [feature/introduce-post-type-glung-mariotest](https://github.com/wordpress-mobile/WordPress-FluxC-Android/tree/feature/introduce-post-type-glung-mariotest)
- WPAndroid repo: [feature/bump-FluxC-version-with-portfolio-glung-mariotest](https://github.com/wordpress-mobile/WordPress-Android/tree/feature/bump-FluxC-version-with-portfolio-glung-mariotest)

(I also changed the references to FluxC repo in the WPAndroid project back to referencing `com.github.wordpress-mobile.WordPress-FluxC-Android:` in 5f4527fb18 in my test branch). I think what was happening is jitpack wasn't catching up with the references on your `glung` repos and it just started to work after the new branches were created (and a first build attempt on Android Studio was made so it tried and triggered the build on jitpack at first request).

**tl;dr** it should work on your branch as well now, if you change the reference back to `com.github.wordpress-mobile.WordPress-FluxC-Android:`.
If you need to make changes to the FluxC branch, then you might probably as well just work on my test branch (and open a new PR from there) so you can get the commits available as you work on it.

However, when running the branch the app crashes upon opening with this log:

```
09-03 15:10:52.587 1256-1256/org.wordpress.android D/AndroidRuntime: Shutting down VM
09-03 15:10:52.599 1256-1256/org.wordpress.android E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.wordpress.android, PID: 1256
    java.lang.RuntimeException: Unable to create application org.wordpress.android.WordPressDebug: java.lang.IllegalStateException: Foreign Key Constraints cannot be enabled or disabled while there are transactions in progress.  Finish all transactions and release all active database connections first.
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5794)
        at android.app.ActivityThread.-wrap1(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1661)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6541)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
     Caused by: java.lang.IllegalStateException: Foreign Key Constraints cannot be enabled or disabled while there are transactions in progress.  Finish all transactions and release all active database connections first.
        at android.database.sqlite.SQLiteConnectionPool.reconfigure(SQLiteConnectionPool.java:287)
        at android.database.sqlite.SQLiteDatabase.setForeignKeyConstraintsEnabled(SQLiteDatabase.java:1863)
        at org.wordpress.android.fluxc.persistence.WellSqlConfig.setForeignKeyConstraintsDisabled(WellSqlConfig.java:387)
        at org.wordpress.android.fluxc.persistence.WellSqlConfig.onUpgrade(WellSqlConfig.java:66)
        at com.yarolegovich.wellsql.WellSql.onUpgrade(WellSql.java:44)
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:299)
        at android.database.sqlite.SQLiteOpenHelper.getReadableDatabase(SQLiteOpenHelper.java:218)
        at com.yarolegovich.wellsql.WellSql.select(WellSql.java:69)
        at org.wordpress.android.fluxc.persistence.AccountSqlUtils.getAccountByLocalId(AccountSqlUtils.java:107)
        at org.wordpress.android.fluxc.persistence.AccountSqlUtils.getDefaultAccount(AccountSqlUtils.java:98)
        at org.wordpress.android.fluxc.store.AccountStore.loadAccount(AccountStore.java:1153)
        at org.wordpress.android.fluxc.store.AccountStore.<init>(AccountStore.java:754)
        at org.wordpress.android.fluxc.store.AccountStore_Factory.get(AccountStore_Factory.java:61)
        at org.wordpress.android.fluxc.store.AccountStore_Factory.get(AccountStore_Factory.java:14)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at org.wordpress.android.WordPress_MembersInjector.injectMembers(WordPress_MembersInjector.java:100)
        at org.wordpress.android.WordPress_MembersInjector.injectMembers(WordPress_MembersInjector.java:17)
        at org.wordpress.android.modules.DaggerAppComponentDebug.inject(DaggerAppComponentDebug.java:2265)
        at org.wordpress.android.WordPress.onCreate(WordPress.java:223)
        at org.wordpress.android.WordPressDebug.onCreate(WordPressDebug.java:14)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1118)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5791)
        at android.app.ActivityThread.-wrap1(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1661) 
        at android.os.Handler.dispatchMessage(Handler.java:105) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6541) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767) 
09-03 15:10:52.604 1728-2212/system_process W/ActivityManager:   Force finishing activity org.wordpress.android/.ui.WPLaunchActivity

```

I've tried this on 2 different handsets, with a previous version of WPAndroid already installed and logged in. Maybe you can try that and see if you can pinpoint the problem?

I confirm this problem does not happen on a fresh install (i.e. first deleted any pevious version of WPAndroid and then installing).

**Comment by glung** [comment]
Hey @mzorz, I find the issue and fixed it here: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/869/commits/a2034b5e57144a89a797937aadad5b1a2bdb0d93

Unfortunately, I do not have the right push on your branch. Could you cherry pick this change maybe?

In a nutshell, the foreign key constraints can’t be added/removed during the migration phase. This is because the migration happens within a transaction. Thankfully, the foreign key does not appear to be an issue. So I guess, we won't follow the SQLite documentation.

**Comment by mzorz** [comment]
Thanks for jumping in @glung 

> In a nutshell, the foreign key constraints can’t be added/removed during the migration phase. This is because the migration happens within a transaction. Thankfully, the foreign key does not appear to be an issue. So I guess, we won't follow the SQLite documentation.

Understood, this seems the way to go then!

The issue described above seems to be solved now as you pointed out, however after bumping the FluxC hash version to point to your fix in a2034b5e57144a89a797937aadad5b1a2bdb0d93, then performing the following specific test:

1. install the version currently in the `develop` branch
2. log in to WP
3. then switch branches to `feature/bump-FluxC-version-with-portfolio-glung-mariotest` (I've updated it with your fix in FluxC), build and install this version over the one installed in step 1.
4. upon opening, the app crashes and this is what I see in the logs:

```
09-03 20:58:48.342 3747-3747/org.wordpress.android E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.wordpress.android, PID: 3747
    java.lang.RuntimeException: Unable to resume activity {org.wordpress.android/org.wordpress.android.ui.main.WPMainActivity}: java.lang.IllegalStateException: cursor.getString(columnIndex) must not be null
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3581)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3621)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2862)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: java.lang.IllegalStateException: cursor.getString(columnIndex) must not be null
        at org.wordpress.android.models.SiteSettingsModel.getStringFromCursor(SiteSettingsModel.kt:319)
        at org.wordpress.android.models.SiteSettingsModel.deserializeOptionsDatabaseCursor(SiteSettingsModel.kt:172)
        at org.wordpress.android.ui.prefs.SiteSettingsInterface.loadCachedSettings(SiteSettingsInterface.java:1010)
        at org.wordpress.android.ui.prefs.SiteSettingsInterface.init(SiteSettingsInterface.java:959)
        at org.wordpress.android.ui.main.MySiteFragment.initSiteSettings(MySiteFragment.java:243)
        at org.wordpress.android.ui.main.MySiteFragment.onResume(MySiteFragment.java:181)
        at android.support.v4.app.Fragment.performResume(Fragment.java:2390)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1474)
        at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1759)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1827)
        at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3244)
        at android.support.v4.app.FragmentManagerImpl.dispatchResume(FragmentManager.java:3212)
        at android.support.v4.app.FragmentController.dispatchResume(FragmentController.java:217)
        at android.support.v4.app.FragmentActivity.onResumeFragments(FragmentActivity.java:509)
        at android.support.v4.app.FragmentActivity.onPostResume(FragmentActivity.java:498)
        at android.support.v7.app.AppCompatActivity.onPostResume(AppCompatActivity.java:171)
        at android.app.Activity.performResume(Activity.java:7141)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3556)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3621) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2862) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6494) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
09-03 20:58:48.344 1118-2359/? W/ActivityManager:   Force finishing activity org.wordpress.android/.ui.main.WPMainActivity
```

Would you mind merging `develop` back into your branch and performing the actual test steps to see if it happens on your end? Thanks in advance

**Comment by mzorz** [comment]
Closing this PR for now as it's not actively maintained
